hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationManagerUtil.java

```diff
@@ -188,4 +195,105 @@ public List<DatanodeDetails> getUsedNodes() {
+   * are no spare nodes to create new replicas on, due to having too many
+   * unhealthy replicas or quasi-closed replicas which cannot be closed due to
+   * having a lagging sequence ID. The logic here will select a replica to
+   * delete, or return null if there are none which can be safely deleted.
+   * @param containerInfo The container to select a replica to delete from
+   * @param replicas The list of replicas for the container
+   * @param pendingOps The list of pending replica operations for the container
+   * @return A replica to delete, or null if there are none which can be safely
+   *         deleted.
+   */
+  public static ContainerReplica selectUnhealthyReplicaForDelete(
+      ContainerInfo containerInfo, Set<ContainerReplica> replicas,
+      List<ContainerReplicaOp> pendingOps,
+      Function<DatanodeDetails, NodeStatus> nodeStatusFn) {
+
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        // There is at least one pending delete which will free up a node.
+        // Therefore we do nothing until that delete completes or times out.
+        LOG.debug("Container {} has pending deletes which will free nodes",
+            pendingOps);
+        return null;
+      }
+    }
+
+    if (replicas.size() <= 2) {
+      // We never remove replicas if it will leave us with less than 2 replicas
+      LOG.debug("There are only {} replicas for container {} so no more will " +
+          "be deleted", replicas.size(), containerInfo);
+      return null;
+    }
+
+    boolean foundMatchingReplica = false;
+    for (ContainerReplica replica : replicas) {
+      if (compareState(containerInfo.getState(), replica.getState())) {
+        foundMatchingReplica = true;
+        break;
+      }
+    }
+    if (!foundMatchingReplica) {
+      LOG.debug("No matching replica found for container {} with replicas " +
+              "{}. No unhealthy replicas can be safely delete.",
+          containerInfo, replicas);
+      return null;
+    }
+
+    List<ContainerReplica> deleteCandidates = new ArrayList<>();
+    for (ContainerReplica r : replicas) {
+      NodeStatus nodeStatus = nodeStatusFn.apply(r.getDatanodeDetails());
+      if (nodeStatus == null || !nodeStatus.isHealthy()
+          || !nodeStatus.isInService()) {
+        continue;
+      }
+      // If the container is closed, then quasi-closed and unhealthy replicas
+      // will get added to this list. If the container is quasi-closed, then
+      // only unhealthy replicas will get added here.
+      if (!compareState(containerInfo.getState(), r.getState())) {
+        deleteCandidates.add(r);
+      }
+    }
+    deleteCandidates.sort(
+        Comparator.comparingLong(ContainerReplica::getSequenceId));
+    if (containerInfo.getState() == HddsProtos.LifeCycleState.CLOSED) {
+      return deleteCandidates.size() > 0 ? deleteCandidates.get(0) : null;
```

Comment:
We don't want to delete a quasi closed replica with the same seq id as the container. The logic from my comment above or some variation should take care of this.